[HigherOrderOp] Remove _deprecated_global_ns from cond

ydwu4 opened the pull request:
Remove _deprecated_global_ns from cond following #104105.

We change the module attribute of HigherOrderOperator instances in the constructor from torch.ops to torch.ops.higher_order when self.namespace is "higher_order". For subclasses (e.g. customized higher order operator), we leave their \_\_module\_\_ unchanged.

Will import this PR to fix internal tests.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

zou3519 commented on torch/fx/node.py:
```diff
@@ -78,6 +78,9 @@ def _type_repr(obj):
     return repr(obj)
 
 def _get_qualified_name(func: Callable[..., Any]) -> str:
+    # higher_order_operators
+    if isinstance(func, torch._ops.HigherOrderOperator) and func.name() in torch._ops._higher_order_ops:
+        return f"torch.ops.higher_order.{func.name()}"
```
Does this also change `map`? If so, that may be a problem, because map is still torch.ops.map.

The check isn't too robust:
1) Ideally cond's `__name__` is "cond" and `__module__` is "torch.ops.higher_order". (Is this the case?)
2) If so, then we should return `f"{func.__module__}.{func.__name__}`

ydwu4 replied:
This pr doesn't deal with map. My plan is to experiment on cond a bit to see how it goes then do it for map in the next pr.

1. cond's \_\_name\_\_ is "cond" but \_\_module\_\_ is "torch._ops". I can check a bit how to make its \_\_module\_\_ torch.ops.higher_order.
2. Yeah, I agree.

zou3519 commented on torch/_ops.py:
```diff
@@ -768,7 +776,14 @@ def __init__(self, name, ops):
 
     def __getattr__(self, name):
-        return self._ops[name]
+        # Following _OpNamespace.__getattr__, we cache the op on the _PyOpNamespace object.
+        op = self._ops.get(name, None)
+        if op is None:
+            raise AttributeError(
+                f"'_PyOpNamespace' '{self.name}' object has no attribute '{name}'"
+            )
+        setattr(self, name, op)
+        return op
```
Is this necessary for this PR? Or is it just an optimization?

ydwu4 replied:
yeah, it's just an optimization. I was looking at how other opearator's \_\_module\_\_ is set and found this logic and the error message seems pretty good. So I "copied" it over.

zou3519 replied:
Sounds good, thanks for explaining

zou3519 reviewed (approved):
LGTM assuming the optimization isn't necessary for this PR. If it is then I am probably misunderstanding something

ydwu4 committed:
[HigherOrderOp] Remove _deprecated_global_ns from cond (#104380)

Summary:
Remove _deprecated_global_ns from cond following https://github.com/pytorch/pytorch/issues/104105.

To do that, we need to change how graph_module generates python_code for "cond" target. Otherwise, it will generate target as "torch.ops.cond", which is invalid after the change.

Will import this PR to fix internal tests.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

Pull Request resolved: https://github.com/pytorch/pytorch/pull/104380

Reviewed By: zou3519

Differential Revision: D47110919

Pulled By: ydwu4

fbshipit-source-id: e0b8090982cee2c1415abb663bb6308eeb4a52b4
